src/Alley/Internals/Traverser.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+/*
+ * This file is part of the traverse/reshape package.
+ *
+ * (c) Alley <info@alley.co>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Alley\Internals;
+
+/**
+ * Contains the implementation details for the `traverse()` functions.
+ *
+ * Internals are not subject to semantic-versioning constraints.
+ */
+final class Traverser
+{
+    /**
+     * Data source.
+     *
+     * @var array|object
+     */
+    private $source;
+
+    /**
+     * Path or array of paths to data.
+     *
+     * @var array|string
+     */
+    private $path;
+
+    /**
+     * Path delimiter.
+     *
+     * @var string
+     */
+    private string $delimiter;
+
+    /**
+     * Found result.
+     *
+     * @var mixed
+     */
+    private $result = null;
+
+    /**
+     * Constructor.
+     *
+     * @param array|object $source Data source.
+     * @param string|array $path The path or array of paths.
+     * @param string $delimiter Delimiter.
+     */
+    public function __construct($source, $path, string $delimiter)
+    {
+        $this->source = $source;
+        $this->path = $path;
+        $this->delimiter = $delimiter;
+    }
+
+    /**
+     * Create an instance and return the traversed value.
+     *
+     * @param mixed ...$args Constructor arguments.
+     * @return mixed
+     */
+    public static function createAndGet(...$args)
+    {
+        $instance = new self(...$args);
+        return $instance->get();
+    }
+
+    /**
+     * Get the traversed value.
+     *
+     * @return mixed
+     */
+    public function get()
+    {
+        $this->run();
+
+        return $this->result;
+    }
+
+    /**
+     * Run the traversal.
+     */
+    private function run()
+    {
+        if (\is_array($this->path)) {
```

Comment:
I generally prefer using `use` statements up top rather than fully qualified paths in the class body. Additionally, the PHP builtins typically don't require a `use` statement or a fully qualified path (unlike custom functions in WP or other libraries).

Reply:
Built-in functions don't require qualification, true, but they can be made infinitesimally faster in some cases when they're fully qualified: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3995.

It turns out I was using an old version of PHP-CS-Fixer that added namespaces to all built-in functions. Now that I've upgraded and only some functions are namespaced, I think I'll keep away from `use` statements for now, since that would make for an inconsistency: some functions would have `use` statements, some wouldn't.